### PR TITLE
CLOUDSTACK-7958: Add configuration for limit to CIDRs for Admin API calls

### PR DESCRIPTION
**Pull request by wido**
The global setting 'management.admin.cidr' is set to 0.0.0.0/0,::/0
by default preserve the current behavior and thus allow API calls
for Admin accounts from all IPv4 and IPv6 subnets.

Users can set it to a comma-separated list of IPv4/IPv6 subnets to
restrict API calls for Admin accounts to certain parts of their network(s).

This is to improve Security. Should a attacker steal the Access/Secret key
of a Admin account he/she still needs to be in a subnet from where Admin accounts
are allowed to perform API calls.

This is a good security measure for APIs which are connected to the public internet.

This PR also includes a commit to cleanup and improve NetUtils.

No existing methods have been altered. That has been verified by adding additional Unit Tests for this.

### CODE REVIEW

File: core/src/com/cloud/network/HAProxyConfigurator.java [hunk 1]
```diff
@@ -538,7 +536,7 @@ private String getLbSubRuleForStickiness(final LoadBalancerTO lbTO) {
         if (stickinessSubRule != null && !destsAvailable) {
             s_logger.warn("Haproxy stickiness policy for lb rule: " + lbTO.getSrcIp() + ":" + lbTO.getSrcPort() + ": Not Applied, cause:  backends are unavailable");
         }
-        if (publicPort.equals(NetUtils.HTTP_PORT) && !keepAliveEnabled || httpbasedStickiness) {
+        if (publicPort == NetUtils.HTTP_PORT && !keepAliveEnabled || httpbasedStickiness) {
```
Comment (DaanHoogland) [inline core/src/com/cloud/network/HAProxyConfigurator.java]: why this? equals() seems more what is intended then ==
is making it an int really the best? as opposed to Integer (or String for that matter)

File: api/src/org/apache/cloudstack/config/ApiServiceConfiguration.java [hunk 2]
```diff
@@ -28,6 +28,8 @@
             "API end point. Can be used by CS components/services deployed remotely, for sending CS API requests", true);
     public static final ConfigKey<Long> DefaultUIPageSize = new ConfigKey<Long>("Advanced", Long.class, "default.ui.page.size", "20",
             "The default pagesize to be used by UI and other clients when making list* API calls", true, ConfigKey.Scope.Global);
+    public static final ConfigKey<String> ManagementAdminCidr = new ConfigKey<String>("Advanced", String.class, "management.admin.cidr",
+            "0.0.0.0/0,::/0", "Comma separated list of IPv4/IPv6 CIDRs from which Admin accounts can perform API calls", true, ConfigKey.Scope.Global);
```
Comment (DaanHoogland) [inline api/src/org/apache/cloudstack/config/ApiServiceConfiguration.java]: argument for this default: backwards compatible
argument against: inherent security risk

File: server/src/com/cloud/api/ApiServlet.java [hunk 3]
```diff
@@ -290,6 +292,17 @@ void processRequestInContext(final HttpServletRequest req, final HttpServletResp
                 CallContext.register(accountMgr.getSystemUser(), accountMgr.getSystemAccount());
             }
 
+            if (CallContext.current().getCallingAccount().getType() == Account.ACCOUNT_TYPE_ADMIN) {
+                s_logger.debug("CIDRs from which Admin accounts are allowed to perform API calls " + adminCidrs);
```
Comment (DaanHoogland) [inline server/src/com/cloud/api/ApiServlet.java]: you don't want to log this on every call.

File: server/src/com/cloud/api/ApiServlet.java [hunk 4]
```diff
@@ -290,6 +292,17 @@ void processRequestInContext(final HttpServletRequest req, final HttpServletResp
...
+                if (!NetUtils.isIpInCidrList(InetAddress.getByName(remoteAddress), adminCidrs.split(","))) {
+                    auditTrailSb.append(" " + HttpServletResponse.SC_UNAUTHORIZED + " " + "IP-Address of remote not in configured Admin CIDR list");
+                    final String serializedResponse =
+                            apiServer.getSerializedApiError(HttpServletResponse.SC_UNAUTHORIZED, "IP-Address of remote not in configured Admin CIDR list",
+                                    params, responseType);
+                    HttpUtils.writeHttpResponse(resp, serializedResponse, HttpServletResponse.SC_UNAUTHORIZED, responseType, apiServer.getJSONContentType());
```
Comment (DaanHoogland) [inline server/src/com/cloud/api/ApiServlet.java]: this you do want to log on every attempt (WARN or INFO???)

File: core/src/com/cloud/network/HAProxyConfigurator.java (hunk 1 above)
Comment (wido) [inline core/src/com/cloud/network/HAProxyConfigurator.java]: It is a port number so it should be a Int. It's a part of the NetUtils refactor. I found that only HTTP/HTTPS port were Strings.

I wanted to make the change in other files as small as possible.

File: api/src/org/apache/cloudstack/config/ApiServiceConfiguration.java (hunk 2 above)
Comment (wido) [inline api/src/org/apache/cloudstack/config/ApiServiceConfiguration.java]: I agree. I have set it to open for all for now. We can submit a different PR afterwards to change the default imho.

File: server/src/com/cloud/api/ApiServlet.java (hunk 3 above)
Comment (wido) [inline server/src/com/cloud/api/ApiServlet.java]: You suggest setting this to TRACE instead of debug?

File: server/src/com/cloud/api/ApiServlet.java (hunk 4 above)
Comment (wido) [inline server/src/com/cloud/api/ApiServlet.java]: True, true. I would say WARN?

File: server/src/com/cloud/api/ApiServlet.java (hunk 3 above)
Comment (DaanHoogland) [inline server/src/com/cloud/api/ApiServlet.java]: I was thinking trace here and debug or info on the first load of the cidr

**Comment by wido** [comment]
@DaanHoogland: I improved the logging as you suggested/requested.

A TRACE for every request and WARN when a request is denied. Tried this locally:

<pre>
2017-04-14 15:45:58,901 WARN  [c.c.a.ApiServlet] (catalina-exec-17:ctx-5955fcab ctx-c572b42e) (logid:7b251506) Request by accountId 2 was denied since 192.168.122.1 does not match 127.0.0.1/8,::1/128
<pre>

In this case only localhost (IPv4/IPv6) is allowed to perform requests.

**Comment by wido** [comment]
@PaulAngus This is what we talked about in Prague. Mind taking a look?

**Comment by remibergsma** [comment]
Nice @wido, will give it a go soon!

File: server/src/com/cloud/api/ApiServlet.java [hunk 5]
```diff
@@ -290,6 +292,19 @@ void processRequestInContext(final HttpServletRequest req, final HttpServletResp
...
+            if (CallContext.current().getCallingAccount().getType() == Account.ACCOUNT_TYPE_ADMIN) {
+                s_logger.trace("CIDRs from which Admin accounts are allowed to perform API calls: " + adminCidrs);
+                if (!NetUtils.isIpInCidrList(InetAddress.getByName(remoteAddress), adminCidrs.split(","))) {
+                    auditTrailSb.append(" " + HttpServletResponse.SC_UNAUTHORIZED + " " + "IP-Address of remote not in configured Admin CIDR list");
+                    s_logger.warn("Request by accountId " + CallContext.current().getCallingAccount().getId() +
+                                  " was denied since " + remoteAddress + " does not match " + adminCidrs);
+                    final String serializedResponse =
+                            apiServer.getSerializedApiError(HttpServletResponse.SC_UNAUTHORIZED, "IP-Address of remote not in configured Admin CIDR list",
+                                    params, responseType);
+                    HttpUtils.writeHttpResponse(resp, serializedResponse, HttpServletResponse.SC_UNAUTHORIZED, responseType, apiServer.getJSONContentType());
```
Comment (remibergsma) [inline server/src/com/cloud/api/ApiServlet.java]: @wido Great work, works fine! I'd recommend adding a `return;` line after the `writeHttpResponse` on line 304 because otherwise it will generate invalid json:

```
Error processing json: {"listvirtualmachinesresponse":{"uuidList":[],"errorcode":401,"errortext":"IP-Address of remote not in configured Admin CIDR list"}}{"listvirtualmachinesresponse":{}}
```

With `return;` added it works as expected:

image

Regards, Remi

**Comment by wido** [comment]
Thanks @remibergsma. Added the return statement.

Thinking about it. Does a 401 sound good? Or should we maybe use a 403 Forbidden?

**Comment by remibergsma** [comment]
@wido I'm playing a bit with it, also because currently CloudMonkey displays a clear message but the UI will simply freeze and do nothing. Only when you look at the underlying API calls you'll see why it isn't working. I'm testing to see the http status code makes any difference of that we need to handle it in the UI somewhere.

We should also think about the order: first alert on the CIDR check and then check user/pass (as it is now) or the other way around.

Also noticed it doesn't work with spaces before/after comma's so we might want to add a `.replaceAll("\\s","")` or similar.

**Comment by ustcweizhou** [comment]
@wido very nice feature.
It would also be nice if this can be added to domain or account setting.

**Comment by wido** [comment]
@remibergsma: Yes, I am aware of that UI problem. Not sure how to fix it.

After thinking about it, I went for '403 Forbidden' and also stripped whitespace from the config key.

I think the order is OK right now unless other opinions?

@ustcweizhou: That is a lot more difficult then a global value, isn't it? Since you have to query it every time.

Or should configkey allow this very easily?

**Comment by remibergsma** [comment]
@wido I think we need to do the check on two places, also on the login() method. That makes sure we don't issue a session key when user/pass are OK but we still reject it based on the CIDR. In my testing that also fixes the UI issue. There are two ways to authenticate so that makes sense I'd say. It'll then also work with authentication plugins, such as LDAP/AD.

Switching the scope of the config is easy, but indeed you'll be querying it on every API call. That does have the benefit you don't need to restart the mgt server when you make a change, but the downside is also obvious. One way to resolve it, is to make a global config setting that switches the feature on/off (and that config is loaded at bootstrap) so you can opt-in for the more heavy checks.

I'll play a bit more with it tonight.

**Comment by wido** [comment]
@remibergsma: I pulled your code, thanks! It now works per account @ustcweizhou 

How does this look?

File: server/src/com/cloud/user/AccountManagerImpl.java [hunk 6]
```diff
@@ -2011,102 +2009,11 @@ public void logoutUser(long userId) {
     }
 
     @Override
-    public UserAccount authenticateUser(String username, String password, Long domainId, InetAddress loginIpAddress, Map<String, Object[]> requestParameters) {
+    public UserAccount authenticateUser(final String username, final String password, final Long domainId, final InetAddress loginIpAddress, final Map<String, Object[]>
+            requestParameters) {
         UserAccount user = null;
         if (password != null && !password.isEmpty()) {
             user = getUserAccount(username, password, domainId, requestParameters);
-        } else {
```
Comment (remibergsma) [inline server/src/com/cloud/user/AccountManagerImpl.java]: @wido You may want to put back this `else` statement. In Cosmic sso has been removed, and when I came across this left-over I removed it. However, I think it shouldn't removed from CloudStack to not break sso.

**Comment by wido** [comment]
Good one @remibergsma. I reverted that piece and also the baremetal refusal of users.

**Comment by wido** [comment]
There were some conflicts after changes were made in master. Fixed those.

As this one is merge ready, can it go into master now?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@wido can you check/fix build failures?

**Comment by wido** [comment]
@rhtyd: I went through the logs, but there is nothing that I can find that points to this PR.

What I do see:

<pre>2017-07-19 15:16:07,845 DEBUG [o.a.c.s.i.BaseImageStoreDriverImpl] (RemoteHostEndPoint-1:ctx-09893152) (logid:21cb76f0) Performing image store createTemplate async callback
2017-07-19 15:16:07,859 WARN  [o.a.c.alerts] (RemoteHostEndPoint-1:ctx-09893152) (logid:21cb76f0)  alertType:: 28 // dataCenterId:: 1 // podId:: null // clusterId:: null // message:: Failed to register template: d8e68e73-fa32-4309-97e5-436da98c7568 with error:  HTTP Server returned 403 (expected 200 OK) 
2017-07-19 15:16:07,863 ERROR [o.a.c.s.i.BaseImageStoreDriverImpl] (RemoteHostEndPoint-1:ctx-09893152) (logid:21cb76f0) Failed to register template: d8e68e73-fa32-4309-97e5-436da98c7568 with error:  HTTP Server returned 403 (expected 200 OK)</pre>

Looking at the logs even further:

<pre>
2017-07-19 17:07:59,887 DEBUG [c.c.a.ApiServer] (qtp411594792-25:ctx-019cac6d ctx-c78c28dc) (logid:92eb3c37) CIDRs from which account 'Acct[b707c234-6ca4-11e7-849c-00163e174325-admin]' is allowed to perform API calls: 0.0.0.0/0,::/0
2017-07-19 17:07:59,920 DEBUG [c.c.a.ApiServer] (qtp411594792-24:ctx-65cfefc0 ctx-264d5734) (logid:a06700ca) CIDRs from which account 'Acct[b707c234-6ca4-11e7-849c-00163e174325-admin]' is allowed to perform API calls: 0.0.0.0/0,::/0
2017-07-19 17:08:59,944 DEBUG [c.c.a.ApiServer] (qtp411594792-29:ctx-b8e61282 ctx-e1ad15c2) (logid:54c2e421) CIDRs from which account 'Acct[b707c234-6ca4-11e7-849c-00163e174325-admin]' is allowed to perform API calls: 0.0.0.0/0,::/0
2017-07-19 17:08:59,970 DEBUG [c.c.a.ApiServer] (qtp411594792-25:ctx-3901b66d ctx-5861a4b0) (logid:220067e9) CIDRs from which account 'Acct[b707c234-6ca4-11e7-849c-00163e174325-admin]' is allowed to perform API calls: 0.0.0.0/0,::/0
2017-07-19 17:09:00,840 DEBUG [c.c.a.ApiServer] (qtp411594792-24:ctx-b991e50f ctx-df021a2a) (logid:f3c648cf) CIDRs from which account 'Acct[b707c234-6ca4-11e7-849c-00163e174325-admin]' is allowed to perform API calls: 0.0.0.0/0,::/0
</pre>

Seems like this PR isn't blocking any API-calls or such. Everything seems to pass.

**Comment by rhtyd** [comment]
@wido it's a build failure issue, please see travis (job#1) failure:
```
[[1;34mINFO[m] Compiling 45 source files to /home/travis/build/apache/cloudstack/vmware-base/target/classes
[[1;34mINFO[m] -------------------------------------------------------------
[[1;31mERROR[m] COMPILATION ERROR : 
[[1;34mINFO[m] -------------------------------------------------------------
[[1;31mERROR[m] /home/travis/build/apache/cloudstack/vmware-base/src/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:[1456,20] error: cannot find symbol
[[1;34mINFO[m] 1 error
...
[[1;34mINFO[m] Apache CloudStack VMware Base ...................... [1;31mFAILURE[m [  4.785 s]
```

**Comment by wido** [comment]
Fixed @rhtyd 

<pre>
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 04:52 min (Wall Clock)
[INFO] Finished at: 2017-07-24T14:07:11+02:00
[INFO] Final Memory: 118M/1939M
[INFO] ------------------------------------------------------------------------
</pre>

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@wido still failing, please do a clean rebuild, check Travis?

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 1006
 Hypervisor xenserver
 NetworkType Advanced
 Passed=102
 Failed=11
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_scale_vm.py

 * ContextSuite context=TestScaleVm>:setup Failing since 33 runs

* test_loadbalance.py

 * test_01_create_lb_rule_src_nat Failed

 * test_02_create_lb_rule_non_nat Failed

* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failed

* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 63 runs

* test_volumes.py

 * test_06_download_detached_volume Failing since 3 runs

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 63 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 96 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 96 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 94 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 94 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by wido** [comment]
I tried again @rhtyd 

<pre>
wido@wido-laptop:~/repos/cloudstack$ mvn -T2C clean install
</pre>

This works! All unit tests also pass. No build failure either:

<pre>
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 06:30 min (Wall Clock)
[INFO] Finished at: 2017-07-25T09:27:50+02:00
[INFO] Final Memory: 119M/1556M
[INFO] ------------------------------------------------------------------------
</pre>

**Comment by rhtyd** [comment]
@wido make sure you're not breaking noredist builds as well, see Travis and try to get it green;

```
[[1;34mINFO[m] [1m--- [0;32mmaven-compiler-plugin:3.2:compile[m [1m(default-compile)[m @ [36mcloud-plugin-hypervisor-vmware[0;1m ---[m
[[1;34mINFO[m] Changes detected - recompiling the module!
[[1;34mINFO[m] Compiling 51 source files to /home/travis/build/apache/cloudstack/plugins/hypervisors/vmware/target/classes
[[1;34mINFO[m] -------------------------------------------------------------
[[1;31mERROR[m] COMPILATION ERROR : 
[[1;34mINFO[m] -------------------------------------------------------------
[[1;31mERROR[m] /home/travis/build/apache/cloudstack/plugins/hypervisors/vmware/src/com/cloud/hypervisor/guru/VMwareGuru.java:[534,35] error: cannot find symbol
[[1;34mINFO[m] 1 error
```

**Comment by wido** [comment]
Aha @rhtyd!

I've fixed that

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 1014
 Hypervisor xenserver
 NetworkType Advanced
 Passed=102
 Failed=9
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_vm_snapshots.py

 * test_change_service_offering_for_vm_with_snapshots Failed

* test_deploy_vm_iso.py

 * test_deploy_vm_from_iso Failing since 69 runs

* test_list_ids_parameter.py

 * ContextSuite context=TestListIdsParams>:setup Failing since 45 runs

* test_volumes.py

 * test_06_download_detached_volume Failed

* test_vm_life_cycle.py

 * test_10_attachAndDetach_iso Failing since 69 runs

* test_routers_network_ops.py

 * test_01_isolate_network_FW_PF_default_routes_egress_true Failing since 102 runs

 * test_02_isolate_network_FW_PF_default_routes_egress_false Failing since 102 runs

 * test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true Failing since 100 runs

 * test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false Failing since 100 runs


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_disk_offerings.py

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
I've tried several times now, the environments fails to come up; perhaps @borisstoyanov  can takeover testing...

```
21:38:41 FAILED - RETRYING: TASK: get wait for state of system VMs to be Running (1 retries left).
21:38:47 fatal: [pr2046-t1277-kvm-centos7-mgmt1]: FAILED! => {"attempts": 200, "changed": true, "cmd": "cloudmonkey list systemvms | jq '.systemvm[]| select(.systemvmtype==\"consoleproxy\")|.state'", "delta": "0:00:00.186658", "end": "2017-07-28 20:38:19.617884", "failed": true, "rc": 0, "start": "2017-07-28 20:38:19.431226", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}
21:38:47 
21:38:47 cmd: cloudmonkey list systemvms | jq '.systemvm[]| select(.systemvmtype=="consoleproxy")|.state'
21:38:47 
21:38:47 start: 2017-07-28 20:38:19.431226
```

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@wido I've tried 4 times now, the trillian environment fails to come up due to systemvms fail to come up. Do you think there are changes that might affect systemvm agents/patching? /cc @borisstoyanov

**Comment by borisstoyanov** [comment]
FYI, we haven't seen this in other master PRs, so could be related to these changes...

**Comment by rhtyd** [comment]
@wido can you fix the conflicts?

**Comment by wido** [comment]
@rhtyd Fixed!

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by wido** [comment]
I will fix fhe conflicts asap

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by ustcweizhou** [comment]
very good !

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@wido can you resolve the conflicts, thanks.

**Comment by wido** [comment]
I will in a few days.

Can we merge this one afterwards? I keep resolving conflicts which happen since other code is merged ;)

**Comment by rhtyd** [comment]
Sure, ping me @wido

**Comment by rhtyd** [comment]
Ping @wido - can you fix the conflicts?

**Comment by wido** [comment]
Done!

I'll be back at work next week

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: packaging/systemd/cloudstack-agent.service [hunk 7]
```diff
@@ -18,8 +18,8 @@
 [Unit]
 Description=CloudStack Agent
 Documentation=http://www.cloudstack.org/
-Requires=libvirtd.service
-After=libvirtd.service
```
Comment (rhtyd) [inline packaging/systemd/cloudstack-agent.service]: @wido this needs to be reverted, as the renaming of libvirtd to libvirt-bin happens via the debian packging scripts. This will cause regressions for KVM. If found KVM hosts failing to start.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Merging this based on 2 code reviews lgtm and test results.

**Comment by ravening** [comment]
@wido sorry to bring up the old pr but how can I configure under this account level?
I logged as in a regular user with "Domain admin" role but i dont see any settings tab under the account

Screenshot 2021-12-22 at 12 32 02

**Comment by weizhouapache** [comment]
> @wido sorry to bring up the old pr but how can I configure under this account level? I logged as in a regular user with "Domain admin" role but i dont see any settings tab under the account
> 
> Screenshot 2021-12-22 at 12 32 02

@ravening please have a look at #4339
